src/generated/types.d.ts

```diff
@@ -1020,82 +1267,101 @@ export type PrivacyWebsitesTrackingProtectionModeOption =
-  onProxyError: SinonEventStub;
   ProxyConfig: {
-    proxyType?: string;
+    autoConfigUrl?: string;
+    autoLogin?: boolean;
+    ftp?: string;
     http?: string;
     httpProxyAll?: boolean;
-    ftp?: string;
-    ssl?: string;
-    socks?: string;
-    socksVersion?: number;
     passthrough?: string;
-    autoConfigUrl?: string;
-    autoLogin?: boolean;
     proxyDNS?: boolean;
+    proxyType?: 'none' | 'autoDetect' | 'system' | 'manual' | 'autoConfig';
+    socks?: string;
+    socksVersion?: number;
+    ssl?: string;
   };
+  onError: Events['Event'];
+  onProxyError: Events['Event'];
+  onRequest: Events['Event'];
+  register: sinon.SinonStub;
+  registerProxyScript: sinon.SinonStub;
   settings: Types['Setting'];
+  unregister: sinon.SinonStub;
 }
 
 export interface Runtime {
-  getBackgroundPage: sinon.SinonStub;
-  openOptionsPage: sinon.SinonStub;
-  getManifest: sinon.SinonStub;
-  getURL: sinon.SinonStub;
-  setUninstallURL: sinon.SinonStub;
-  reload: sinon.SinonStub;
-  requestUpdateCheck?: sinon.SinonStub;
-  restart?: sinon.SinonStub;
-  connect: sinon.SinonStub;
-  connectNative: sinon.SinonStub;
-  sendMessage: sinon.SinonStub;
-  sendNativeMessage: sinon.SinonStub;
-  getBrowserInfo: sinon.SinonStub;
-  getPlatformInfo: sinon.SinonStub;
-  getPackageDirectoryEntry?: sinon.SinonStub;
-  onStartup: SinonEventStub;
-  onInstalled: SinonEventStub;
-  onSuspend?: SinonEventStub;
-  onSuspendCanceled?: SinonEventStub;
-  onUpdateAvailable: SinonEventStub;
-  onBrowserUpdateAvailable?: SinonEventStub;
-  onConnect: SinonEventStub;
-  onConnectExternal: SinonEventStub;
-  onMessage: SinonEventStub;
-  onMessageExternal: SinonEventStub;
-  onRestartRequired?: SinonEventStub;
-  Port: {
+  BrowserInfo: {
+    buildID: string;
     name: string;
+    vendor: string;
+    version: string;
   };
   MessageSender: {
     frameId?: number;
     id?: string;
-    url?: string;
+    tab?: Tabs['Tab'];
     tlsChannelId?: string;
+    url?: string;
   };
-  PlatformOs: RuntimePlatformOs[];
-  PlatformArch: RuntimePlatformArch[];
-  PlatformInfo: {};
-  BrowserInfo: {
-    name: string;
-    vendor: string;
-    version: string;
-    buildID: string;
+  OnInstalledReason: RuntimeOnInstalledReason;
+  OnRestartRequiredReason: RuntimeOnRestartRequiredReason;
+  PlatformArch: RuntimePlatformArch;
+  PlatformInfo: {
+    arch: Runtime['PlatformArch'];
+    nacl_arch?: any;
+    os: Runtime['PlatformOs'];
   };
-  RequestUpdateCheckStatus: RuntimeRequestUpdateCheckStatus[];
-  OnInstalledReason: RuntimeOnInstalledReason[];
-  OnRestartRequiredReason: RuntimeOnRestartRequiredReason[];
-  lastError: {
+  PlatformOs: RuntimePlatformOs;
+  Port: {
+    disconnect: sinon.SinonStub;
+    name: string;
+    onDisconnect: Events['Event'];
+    onMessage: Events['Event'];
+    postMessage: sinon.SinonStub;
+    sender?: Runtime['MessageSender'];
+  };
+  RequestUpdateCheckStatus: RuntimeRequestUpdateCheckStatus;
+  connect: () => Runtime['Port'];
```

Comment:
I'd expect this to be an sinon stub interface that also returns the Runtime['Port'], so something like

```ts
interface SinonStubReturns<T> extends sinon.SinonStub {
  () => T;
}

interface Runtime {
  connect: SinonStubReturns<Runtime['Port']>
}
```

That allows asserting calls to connect(), and I believe it's what the stub generator already does?

Reply:
Ok sure, sounds good. I'm pretty unfamiliar with `Sinon`, so I definitely welcome your help / guidance regarding what `Sinon` types to use. Yes the `stubs` generator calls `stub().returns(returnValue)`. But I wasn't sure how to convert this to a typescript `interface` property. I also couldn't find where `sinon.SinonStub` is defined. I've looked in `node_modules/sinon` but can't find any definitions for `SinonStub`. Any ideas?

Regarding your `SinonStubReturns` interface - does that also already exist? Or would we need to add it to the `types.d.ts` file?

Reply:
The [sinon types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/sinon/ts3.1/index.d.ts) come from [`@types/sinon`](https://github.com/stoically/webextensions-api-mock/blob/fa55f0e235c480ecf6ab1f0eeddceb00848956f6/package.json#L49). 

> Regarding your SinonStubReturns interface - does that also already exist? Or would we need to add it to the types.d.ts file?

That doesn't exist, and adding it like that should be sufficient, yeah.

Reply:
Aha, I've found it: `node_modules/\@types/sinon/index.d.ts`

Reply:
Ok I've now changed this in https://github.com/stoically/webextensions-api-mock/pull/4/commits/b0058ba823df856f0470e41c9e8dac287c5aa195 as follows:

```typescript
interface Runtime {
    connect: sinon.SinonStub<any[], Runtime['Port']>;
}
```

This is based on the types I found in `node_modules/\@types/sinon/ts3.1/index.d.ts`. And it seems to build ok!